source/Optimizers/nelder_mead.d

```diff
@@ -0,0 +1,519 @@
+                 T contraction_coef=0.5,
+                 T shrink_coef=0.5) {
+    return nelder_mead!T(_v, toDelegate(_func),
+                         reflection_coef, expension_coef,
+                         contraction_coef, shrink_coef);
+}
+
+T nelder_mead(T)(ref Vector!T _v, T delegate(in Vector!T) _func,
+                 in T _random_bound=5.0,
+                 in size_t _max_iterations=10000,
+                 in T _reflection_coef=1.0,
+                 in T _expension_coef=2.0,
+                 in T _contraction_coef=0.5,
+                 in T _shrink_coef=0.5)
+{
+    /+
+     + Arguments:
+     +
+     +  - _v (Vector!T): will contains a local minimum of the _func.
+     +
+     +  - _func (T delegate(Vector!T)): the function to minimize.
+     +
+     +  - _random_bound (T): Absolute upper/lower bound for the creation
+     +                       of the vectors.
+     +
+     +  - _max_iterations (size_t): Maximum number of iterations.
+     +
+     +  - reflection_coef (T): Meta-Parameter.
+     +
+     +  - expension_coef (T): Meta-Parameter.
+     +
+     +  - contraction_coef (T): Meta-Parameter.
+     +
+     +  - shrink_coef (T): Meta-Parameter.
+     +/
+
+    assert(_contraction_coef <= 0.5, "Contraction coef must be <= 0.5");
+    assert(_contraction_coef > 0, "Contraction coef must be > 0");
+
+    immutable size_t length = _v.length;
+    immutable size_t num_points = _v.length + 1;
+
+    /// Create Simplex
+
+    auto simplex = new Vector!T[num_points];
+    /+
+    foreach(i; 0 .. (num_points))
+        simplex[i] = new Vector!T(length, _random_bound);
+    +/
+    foreach(i; 0 .. (num_points-1)) {
+        simplex[i] = new Vector!T(length, 0);
+        simplex[i].v[i] = _random_bound;
+    }
+    simplex[num_points-1] = new Vector!T(length, 0);
+    foreach(i; 0 .. length)
+        simplex[num_points-1].v[i] = - _random_bound / sqrt(to!real(length));
+
+    /// Compute values at simplex points.
+    T[] simplex_values = new T[num_points];
+    foreach(i; 0 .. num_points)
+        simplex_values[i] = _func(simplex[i]);
+
+    /// Order simplex
+    auto min_value = simplex_values.minElement;
+    auto min_index = simplex_values.minIndex;
+
+    auto tmp_max_value = new T[2];
+    simplex_values.topNCopy!"a > b"(tmp_max_value, Yes.sortOutput);
+    size_t[] index_max = new size_t[2];
+    simplex_values.topNIndex!"a > b"(index_max, Yes.sortOutput);
+    auto max1_value = tmp_max_value[0];
+    auto max1_index = index_max[0];
+    auto max2_value = tmp_max_value[1];
+    auto max2_index = index_max[1];
+            
+    /// Init util vector
+    auto temporary_vector = new Vector!T(length, _random_bound);
+    auto reflection_vector = new Vector!T(length, _random_bound);
+    auto expension_vector = new Vector!T(length, _random_bound);
+    auto contraction_vector = new Vector!T(length, _random_bound);
+    auto shrink_vector = new Vector!T(length, _random_bound);
+    T temporary_value;

```

Comment:
might want to assert T has a default constructor at function beggining? Might be undefined behavior for weird T's here

Reply:
Bien vu